photosynthesis_metrics/psnr.py

```diff
@@ -0,0 +1,53 @@
+r""" This module implements Peak Signal-to-Noise Ration (PSNR) in PyTorch.
+"""
+import torch
+from typing import Optional, Union
+
+from photosynthesis_metrics.utils import _validate_input, _adjust_dimensions
+
+
+def psnr(x: torch.Tensor, y: torch.Tensor, data_range: Union[int, float] = 1.0,
+         reduction: Optional[str] = 'mean', convert_to_greyscale: bool = False):
+    r"""Compute Peak Signal-to-Noise Ration for a batch of images.
+    Supports both greyscale and color images with RGB channel order.
+
+    Args:
+        x: Batch of predicted images with shape (batch_size x channels x H x W)
+        y: Batch of target images with shape  (batch_size x channels x H x W)
+        data_range: Value range of input images (usually 1.0 or 255). Default: 1.0
+        reduction: Reduction over samples in batch: "mean"|"sum"|"none"
+        convert_to_greyscale: Convert RGB image to YCbCr format and computes PSNR
+            only on luminance channel if `True`. Compute on all 3 channels otherwise.
+        
+    Returns:
+        PSNR: Index of similarity betwen two images.
+    Note:
+        Implementaition is based on Wikepedia https://en.wikipedia.org/wiki/Peak_signal-to-noise_ratio
+    """
+    _validate_input((x, y), allow_5d=False)
+    x, y = _adjust_dimensions(input_tensors=(x, y))
+
+    # Constant for numerical stability
+    EPS = 1e-8
+
+    x = x / data_range
+    y = y / data_range
+
+    if (x.size(1) == 3) and convert_to_greyscale:
+        # Convert RGB image to YCbCr and take luminance: Y = 0.299 R + 0.587 G + 0.114 B
+        x = 0.299 * x[:, 0, :, :] + 0.587 * x[:, 1, :, :] + 0.114 * x[:, 2, :, :]
+        y = 0.299 * y[:, 0, :, :] + 0.587 * y[:, 1, :, :] + 0.114 * y[:, 2, :, :]
+
+        # Add channel dimension
+        x = x[:, None, :, :]
+        y = y[:, None, :, :]
```

Comment:
Consider this approach to avoid slicing.
`if x.size(1) == 3:
        # rgb_to_grey - weights to transform RGB image to grey
        rgb_to_grey = torch.tensor([0.299, 0.587, 0.114]).view(1, -1, 1, 1).to(x)
        x = torch.sum(x * rgb_to_grey, dim=1, keepdim=True)
`

Reply:
Good point. Added this